### PR TITLE
CuPy JIT: Readable compile error messages

### PR DESCRIPTION
**Pull request by asi1024**
Indicates on which line of the target Python function the compile error occurred.

If an error is raised while the compilation, the type of error is immediately converted to `_JitCompileError` class temporarily.
The attributes of `_JitCompileError` object include 1. the type and message of the original error and 2. On which line of the original python function the error is raised.
Just before exiting the top level of the compilation function, the error type is restored to the original error type, and "Where in the Python function the compilation error occurred" is shown as the error message.

### CODE REVIEW

File: cupyx/jit/_compile.py [hunk 1]
```diff
@@ -195,6 +227,21 @@ def _transpile_function(
...
+    try:
+        return _transpile_function_internal(
+            func, attributes, mode, consts, in_types, ret_type)
+    except JitCompileError as e:
+        exc = e
+        if _is_debug_mode:
+            exc.reraise(source)
+
+    # Raises the error out of `except` block to clean stack trace.
+    exc.reraise(source)
```
Comment (leofang) [inline cupyx/jit/_compile.py:239]: I am confused by the purpose: Doesn't this mean we always raise the JitCompileError if it happens? Would the stack trace depends on whether we raise it inside the `except` block or not?

File: cupyx/jit/_compile.py [hunk 2]
```diff
@@ -14,11 +14,41 @@
...
 
 
+_is_debug_mode = False
+
 _typeclasses = (bool, numpy.bool_, numbers.Number)
 
 Result = collections.namedtuple('Result', ['func_name', 'code', 'return_type'])
 
 
+class JitCompileError(Exception):
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupyx/jit/_compile.py:24]: This type of error is not exposed to users.

File: cupyx/jit/_compile.py (hunk 1 above)
Comment (asi1024) [reply cupyx/jit/_compile.py:239]: Updated the PR description 😃

Comment (asi1024) [reply cupyx/jit/_compile.py:239]: If reraised the error inside the `except` block, the stack traces of internal functions `_transpile_{stmt/expr}` are also printed, but it is annoying and not benefical for the users. So I implemented to reraise the error outside of `except` block.

Comment (leofang) [reply cupyx/jit/_compile.py:239]: Wow I didn't know this trick, cool!

Before:
```python
Traceback (most recent call last):
  File "test_jit_error.py", line 8, in <module>
    f[1, 1]()
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_interface.py", line 82, in <lambda>
    return lambda *args, **kwargs: self(grid, block, args, **kwargs)
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_interface.py", line 66, in __call__
    _types.Void(),
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 51, in transpile
    tree.body[0], attributes, mode, consts, in_types, ret_type)
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 231, in _transpile_function
    body = _transpile_stmts(func.body, True, env)
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 318, in _transpile_stmts
    codeblocks.extend(_transpile_stmt(stmt, is_toplevel, env))
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 334, in _transpile_stmt
    value = _transpile_expr(stmt.value, env)
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 464, in _transpile_expr
    res = _transpile_expr_internal(expr, env)
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 642, in _transpile_expr_internal
    f'Unbound name: {expr.id} in line {expr.lineno}')
NameError: Unbound name: abc in line 3
```
After:
```python
Traceback (most recent call last):
  File "test_jit_error.py", line 8, in <module>
    f[1, 1]()
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_interface.py", line 82, in <lambda>
    return lambda *args, **kwargs: self(grid, block, args, **kwargs)
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_interface.py", line 66, in __call__
    _types.Void(),
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 82, in transpile
    source=source
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 239, in _transpile_function
    exc.reraise(source)
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 37, in reraise
    raise self.error_type(self.mes + '\n\n' + pycode)
NameError: Unbound name: abc

  @cupyx.jit.rawkernel()
  def f():
>     return abc

```

**Review by emcastillo — APPROVED** [review]
LGTM!

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by asi1024** [comment]
Jenkins, test this please.
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins failure is unrelated to this PR.

**Review by emcastillo — APPROVED** [review]

**Pull request merged**